packages/xo-server/src/xo-mixins/migrate-vm.mjs

```diff
@@ -106,4 +125,153 @@ export default class MigrateVm {
       }
     }
   }
+
+  async migrationfromEsxi({ host, user, password, sslVerify, sr: srId, network: networkId, vm: vmId, thin }) {
+    const esxi = new Esxi(host, user, password, sslVerify)
+    const app = this._app
+
+    await fromEvent(esxi, 'ready')
+    const esxiVmMetadata = await esxi.getTransferableVmMetadata(vmId)
+    const { memory, name_label, networks, numCpu, powerState, snapshots } = esxiVmMetadata
+    const isRunning = powerState !== 'poweredOff'
+
+
+    let chain = []
+    if (snapshots && snapshots.current) {
+      const currentSnapshotId = snapshots.current
+
+      let currentSnapshot = snapshots.snapshots.find(({ uid }) => uid === currentSnapshotId)
+
+      chain = [currentSnapshot.disks]
+      while ((currentSnapshot = snapshots.snapshots.find(({ uid }) => uid === currentSnapshot.parent))) {
+        chain.push(currentSnapshot.disks)
+      }
+      chain.reverse()
+    }
+
+    chain.push(esxiVmMetadata.disks)
+
+    const chainsByNodes = {}
+    chain.forEach(disks => {
+      disks.forEach(disk => {
+        chainsByNodes[disk.node] = chainsByNodes[disk.node] || []
+        chainsByNodes[disk.node].push(disk)
+      })
+    })
+
+    chain[chain.length - 1].forEach(disk => {
+      if (disk.capacity > 2 * 1024 * 1024 * 1024 * 1024) {
+        /* 2TO */
+        throw new Error("Can't migrate disks larger than 2To")
+      }
+    })
+
+    // got data, ready to start creating
+    const sr = app.getXapiObject(srId)
+    const xapi = sr.$xapi
+    const vm = await xapi._getOrWaitObject(
+      await xapi.VM_create({
```

Comment:
Use  `$defer.onFailure` after each creation to roll back in case of error.

See https://github.com/JsCommunity/golike-defer.